### ISSUE TITLE
forge: maybe a test isolation bug with vm.setEnv()

### ISSUE THREAD

**Issue by OliverNChalk** [foundry-rs/foundry#2349]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (37e4376 2022-07-16T00:12:54.96929664Z)

### What command(s) is the bug in?

forge test

### Operating System

Linux

### Describe the bug

Not sure if this is intended behavior or not. When call `vm.setEvn` within a test, it sets the environment for all currently running tests.

This essentially means it is unsafe to run tests that use `env` in parallel. I've tested tests in different files and they  are also not isolated.

Because test are not deterministically scheduled this can also lead to non-deterministic test failures, where one case will race another to set & read the env.

As a reference, this was found while trying to test `solenv`. To address this, all test cases had to be consolidated: https://github.com/proximacapital/solenv/blob/master/test/Solenv.t.sol#L46

Not sure if we are expecting devs to regularly be setting and getting env in tests. I could imagine some integration tests using `setFork` would want to read env and may get tripped by this (if any test sets env)?

**Comment by onbjerg** [comment]
Unsure if this is a bug per se... It does change the external environment, so this is sort of expected? I can't really imagine a case where you would use the same env var with different values in a test run, especially not one where you set the env var using `vm.setEnv`. Re: `vm.setFork` you don't need to use `vm.env` to get that from the environment, you can just use env var interpolation in `foundry.toml`

**Comment by onbjerg** [comment]
Matt did some fixes related to the environment cheatcodes, is this still an issue for you?

**Comment by mattsse** [comment]
to add on that, 
we set the env var of the process that's currently running and since all test are executed in the same process they share the same env values, so there's no isolation.

but since all tests are run in parallel with no deterministic order I wonder if we should just isolate on a per-test basis like:

```
set_vars: HashMap<String, String>
```

but if this is now on a per-test basis this is essentially the same as a test internal solidity var so I think this is a won't fix

**Comment by onbjerg** [comment]
Yeah that's where I was going with my first comment but I wasn't sure if there was a bug I just couldn't spot :smile: Will close as per my first comment and @mattsse's comment
